piptools/scripts/compile.py

```diff
@@ -246,6 +255,14 @@ def cli(
         # Close the file at the end of the context execution
         ctx.call_on_close(safecall(output_file.close_intelligently))
 
+    if index is not None:
+        warnings.warn(
+            "--index and --no-index are deprecated and will be removed "
+            "in future versions. Use --emit-index-url/--no-emit-index-url instead.",
+            category=PipToolsPendingDeprecationWarning,
```

Comment:
Always set stacklevel because you want to point to the end-user's code where they should stop using it, not here.
```suggestion
            category=PipToolsPendingDeprecationWarning,
            stacklevel=2,
```

Reply:
AFAIK it's relevant for libraries, not for cli apps, see the message otherwise with `stacklevel=2`:

```console
$ pip-compile --extra-index-url=https://test.pypi.org/simple/ --no-emit-index-url --index
/Users/albert/Projects/pip-tools/.venv/lib/python3.8/site-packages/click/decorators.py:21: PipToolsPendingDeprecationWarning: --index and --no-index are deprecated and will be removed in future versions. Use --emit-index-url/--no-emit-index-url instead.
  return f(get_current_context(), *args, **kwargs)
#
# This file is autogenerated by pip-compile
# To update, run:
#
#    pip-compile --extra-index-url=https://test.pypi.org/simple/ --index --no-emit-index-url
#
--extra-index-url https://test.pypi.org/simple/

six==1.14.0               # via -r requirements.in
```

Current;y message is:

```
pip-compile --extra-index-url=https://test.pypi.org/simple/ --no-emit-index-url --index
/Users/albert/Projects/pip-tools/piptools/scripts/compile.py:259: PipToolsPendingDeprecationWarning: --index and --no-index are deprecated and will be removed in future versions. Use --emit-index-url/--no-emit-index-url instead.
  warnings.warn(
```

Reply:
Well, maybe you could detect somehow CLI vs import. Can you guarantee that nobody is using the APIs directly via imports?

Reply:
I think nobody uses `piptools.scripts.compile.cli` and calls it because it won't work. AFAIK The only way is to invoke `pip-compile` or `python -m piptools compile` or use click's `CliRunner()`.

Reply:
Kindly ping. What do you think?

Reply:
I'm not concerned too much. Ignore it. After all, done is better than perfect :)

Reply:
Thanks. Nice rule of thumb! 👍